tephra-core/src/main/java/org/apache/tephra/TransactionManager.java

```diff
@@ -1011,10 +1038,9 @@ private boolean doInvalidate(long writePointer) {
       } else {
         // invalidate any checkpoint write pointers
         LongArrayList childWritePointers = previous.getCheckpointWritePointers();
-        if (childWritePointers != null) {
-          for (int i = 0; i < childWritePointers.size(); i++) {
-            invalid.add(childWritePointers.get(i));
-          }
+        for (long childWritePointer : childWritePointers) {
+          invalid.add(childWritePointer);
+          inProgress.remove(childWritePointer);
```

Comment:
Missed this one in last review, we can use the `keySet().removeAll(childWritePointers)` as well (and check if `!childWritePointers.isEmpty()` first). The same for invalid, it can be `invalid.addAll(childWritePointers)`